bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/site/SiteStructureFactory.java

```diff
@@ -65,9 +68,113 @@ public class SiteStructureFactory implements AdapterFactory {
+        Page previewPage = findPreviewPage(xfPage);
+        if (previewPage != null) {
+            return previewPage;
+        }
+
+        // Fallback to progressive path lookup
+        return findCorrespondingPage(xfPage);
+    }
+
+    /**
+     * Recursively tries to find a corresponding page in the main content structure
+     * by progressively removing path segments from the experience fragment path.
+     * 
+     * @param xfPage The experience fragment page
+     * @return The found page or the original page if no match found
+     */
+    private Page findCorrespondingPage(Page xfPage) {
+        String experienceFragmentPath = xfPage.getPath();
+        String relativePath = experienceFragmentPath.substring(EXPERIENCE_FRAGMENTS_PATH.length() + 1);
+        String contentPath = "/content/" + relativePath;
+
+        PageManager pageManager = xfPage.getPageManager();
+        Page foundPage = findCorrespondingPageIterative(pageManager, contentPath);
+
+        return foundPage != null ? foundPage : xfPage;
+    }
+
+    /**
+     * Iteratively tries to find a corresponding page in the main content structure
+     * by progressively removing path segments from the experience fragment path.
+     * 
+     * @param pageManager The page manager to use for page lookups
+     * @param contentPath The full content path (e.g., "/content/venia/us/en/site/header/master")
+     * @return The found page or null if no page exists
+     */
+    private Page findCorrespondingPageIterative(PageManager pageManager, String contentPath) {
+        if (contentPath == null || contentPath.isEmpty()) {
+            return null;
+        }
+
+        // Use iterative approach instead of recursion to avoid stack overflow
+        String currentPath = contentPath;
+
+        while (StringUtils.isNotEmpty(currentPath)) {
+            // Try the current path
+            Page page = pageManager.getPage(currentPath);
+            if (page != null) {
+                return page;
+            }
+
+            // Remove the last segment and try again
+            int lastSlashIndex = currentPath.lastIndexOf('/');
+            if (lastSlashIndex <= 0) {
+                break; // Reached root level
+            }
+
+            currentPath = currentPath.substring(0, lastSlashIndex);
+        }
+
+        return null;
+    }
+
+    /**
+     * Finds the preview page by looking for a node with a non-empty cq:cifPreviewPage attribute.
+     * Traverses up the tree through both pages and folders.
+     * 
+     * @param currentPage The current page to start the search from
+     * @return The preview page if found, otherwise null
+     */
+    private Page findPreviewPage(Page currentPage) {
+        if (currentPage == null) {
+            return null;
+        }
+
+        PageManager pageManager = currentPage.getPageManager();
+
+        // Walk up the tree using resources to support both pages and folders
```

Comment:
Did you consider using https://developer.adobe.com/experience-manager/reference-materials/6-5/javadoc/com/day/cq/commons/inherit/InheritanceValueMap.html it's already used in SiteNavigation.